src/components/Wizard/Patterns/StatefulWizardPattern.js

```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import polyfill from 'react-lifecycles-compat';
+
+import WizardPattern from './WizardPattern';
+import { wizardStepShape } from './WizardPatternConstants';
+import { propOrState, bindMethods } from '../../../index';
+
+/**
+ * StatefulWizardPattern - the Stateful Wizard Pattern component.
+ */
+class StatefulWizardPattern extends React.Component {
+  static getDerivedStateFromProps(nextProps, prevState) {
+    return {
+      activeStepIndex: propOrState(nextProps, prevState, 'activeStepIndex')
+    };
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = { activeStepIndex: 0 };
+    bindMethods(this, ['onStepChanged']);
+  }
+
+  onStepChanged(newStepIndex) {
+    const { shouldPreventStepChange, steps } = this.props;
+    const { activeStepIndex } = this.state;
+    const { shouldPreventExit } = steps[activeStepIndex];
+    const { shouldPreventEnter } = steps[newStepIndex];
+    if (
+      shouldPreventStepChange(activeStepIndex, newStepIndex) ||
+      (shouldPreventExit && shouldPreventExit(newStepIndex)) ||
+      (shouldPreventEnter && shouldPreventEnter(activeStepIndex))
+    )
+      return;
+    this.setState({ activeStepIndex: newStepIndex });
+  }
+
+  render() {
+    const { shouldDisableNextStep, steps, ...otherProps } = this.props;
+    const { activeStepIndex } = this.state;
+    return (
+      <WizardPattern
+        nextStepDisabled={shouldDisableNextStep(activeStepIndex)}
+        {...otherProps}
+        steps={steps.map(step => {
+          const {
+            shouldPreventEnter,
+            shouldPreventExit,
+            ...otherStepProps
+          } = step;
+          return otherStepProps;
```

Comment:
If `StatefulWizardPattern`'s `steps` array element shape is a superset of `WizardPattern`'s `steps` array element shape, do we really need to exclude the extra bits? (i.e. why not simply pass `steps` as-is?)

Reply:
I guess we can. I think I was trying to prevent someone from later on changing WizardPattern to use those props somehow, or wonder why they weren't in the propTypes there. This makes it very explicit that those props are only for StatefulWizardPattern. But it's an unnecessary performance hit, i'll remove it and just leave a comment.

Reply:
If `WizardPattern` doesn't use those extra bits, I think we can pass the steps array to it as-is.

If `WizardPattern` starts to use those extra bits, it'd be an overlap with `StatefulWizardPattern` and we can modify both components as necessary.

If you took this to the extreme, you'd throw an error in `WizardPattern` if any `steps` array element's shape doesn't exactly match the expected structure :smiley: which would obviously be an unnecessary complication.